### PR TITLE
Add support for <optgroup> in SelectField

### PR DESCRIPTION
**Pull request by Giandrop**
Add support for <optgroup> in SelectField as described in issue #656

### CODE REVIEW

**Comment by davidism** [comment]
Please rebase this onto master, 2.3.x will not get another feature release.

File: wtforms/fields/core.py [hunk 1]
```diff
@@ -472,6 +472,12 @@ class SelectField(SelectFieldBase):
     def __init__(self, label=None, validators=None, coerce=text_type, choices=None, validate_choice=True, **kwargs):
         super(SelectField, self).__init__(label, validators, **kwargs)
         self.coerce = coerce
+        if callable(choices):
```
Comment (davidism) [inline wtforms/fields/core.py:475]: Allowing a callable for `choices` is a separate feature request, please make a separate PR.

File: wtforms/fields/core.py [hunk 2]
```diff
@@ -472,6 +472,12 @@ class SelectField(SelectFieldBase):
...
+            choices = choices()
+        if isinstance(choices, dict):
+            self.groups = choices
+        else:
+            self.groups = {}
         self.choices = list(choices) if choices is not None else None
```
Comment (davidism) [inline wtforms/fields/core.py:481]: `self.choices` will not be set to the expected value when a dict is passed in. This still needs to remain consistent (the optgroup doesn't affect the accepted values, only how they are presented).

File: wtforms/fields/core.py [hunk 3]
```diff
@@ -472,6 +472,12 @@ class SelectField(SelectFieldBase):
     def __init__(self, label=None, validators=None, coerce=text_type, choices=None, validate_choice=True, **kwargs):
         super(SelectField, self).__init__(label, validators, **kwargs)
         self.coerce = coerce
+        if callable(choices):
+            choices = choices()
+        if isinstance(choices, dict):
+            self.groups = choices
+        else:
+            self.groups = {}
```
Comment (davidism) [inline wtforms/fields/core.py:480]: Use `None` rather than creating an unused dict.

**Comment by azmeuk** [comment]
Thank you for your contribution. You still have some tests failing, and some style issues.

**Comment by azmeuk** [comment]
@Giandrop are you still interested in solving the issues related to this PR?

**Comment by Giandrop** [comment]
Sorry, I completly forgot about this. I will be revisiting it during this weekend.

**Comment by azmeuk** [comment]
Thank you!
